modules/threads/threads.go

```diff
@@ -0,0 +1,153 @@
+package threads
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/risor-io/risor/internal/arg"
+	"github.com/risor-io/risor/object"
+	"github.com/risor-io/risor/op"
+)
+
+const THREAD object.Type = "thread"
+
+type Thread struct {
+	wg       sync.WaitGroup
+	callable object.Callable
+	args     []object.Object
+	res      object.Object
+	done     bool
+	once     bool
+}
+
+func (t *Thread) Type() object.Type {
+	return THREAD
+}
+
+func (t *Thread) Inspect() string {
+	return "thread"
+}
+
+func (t *Thread) Interface() interface{} {
+	return nil
+}
+
+func (t *Thread) IsTruthy() bool {
+	return t.res != nil
+}
+
+func (t *Thread) Cost() int {
+	return 8
+}
+
+func (t *Thread) MarshalJSON() ([]byte, error) {
+	return nil, fmt.Errorf("type error: unable to marshal %s", THREAD)
+}
+
+func (t *Thread) RunOperation(opType op.BinaryOpType, right object.Object) object.Object {
+	return object.Errorf("eval error: unsupported operation for %s: %v", THREAD, opType)
+}
+
+func (t *Thread) Equals(other object.Object) object.Object {
+	if other.Type() != THREAD {
+		return object.False
+	}
+	return object.False
+}
+
+func (t *Thread) SetAttr(name string, value object.Object) error {
+	switch name {
+	case "once":
+		v, err := object.AsBool(value)
+		if err != nil {
+			return err.Value()
+		}
+		t.done = v
+		return nil
+	}
+	return fmt.Errorf("attribute error: %s object has no attribute %q", THREAD, name)
+}
+
+func (t *Thread) GetAttr(name string) (object.Object, bool) {
+	switch name {
+	case "start", "run":
+		name = "threads." + name
+		return object.NewBuiltin(name, func(ctx context.Context, args ...object.Object) object.Object {
+			if err := arg.Require(name, 0, args); err != nil {
+				return err
+			}
+			if t.once && t.done {
+				if t.res == nil {
+					return object.Nil
+				}
+				return t.res
+			}
+			t.done = true
+			t.wg.Add(1)
+			go func() {
+				defer t.wg.Done()
+				t.res = t.callable.Call(ctx, t.args...)
+			}()
```

Comment:
This approach has a nice charm: it allows the logic to be implemented in a Risor module instead of inside the Risor VM.
It certainly gets a plus on it's simplicity.

However it does not tackle the race condition issue.

Implementing this in the VM could solve that by forcing all Risor code to run on the main thread, and then have the VM switch between "Risor coroutines".

For a scripting language like Risor, I sense that true multi-threaded execution is not really necessary, on the Risor part. Only IO needs to be multi-threaded/concurrent.

Pro's for single-threaded Risor code:
- Never worry about race conditions when writing Risor code.
- No need to expose "thread-safe/concurrent types and utilities", e.g the sync.Map, sync.Once, sync.Mutex, sync.RWMutex, sync.WaitGroup.

Con's:
- No way to do multi-threaded CPU-bound calculations written in Risor.
- Less straight forward to modify the Risor VM.

As the code is running interpreted anyway, I don't see much gain in making Risor code run in parallel, as it's the wrong use case.

Better to try adding good integration with Go so you can offload important calculations to compiled Go code, which in turn could run things in true parallel.

Reply:
Yeah,  this won't work without some changes to the VM but the idea was to provide a simple way of running stuff in the background.